src/ff/galois_field.rs

```diff
@@ -42,6 +42,56 @@ fn assert_copy<C: Copy>(c: C) -> C {
     c
 }
 
+#[cfg(all(target_arch = "x86_64", target_feature = "pclmulqdq"))]
+mod clmul_x86_64 {
+    use core::arch::x86_64::{__m128i, _mm_clmulepi64_si128, _mm_extract_epi64, _mm_set_epi64x};
+
+    unsafe fn clmul(a: u64, b: u64) -> __m128i {
+        #[allow(clippy::cast_possible_wrap)]
+        unsafe fn to_m128i(v: u64) -> __m128i {
+            _mm_set_epi64x(0, v as i64)
+        }
+        _mm_clmulepi64_si128(to_m128i(a), to_m128i(b), 0)
+    }
+
+    #[allow(clippy::cast_sign_loss)]
+    unsafe fn extract<const I: i32>(v: __m128i) -> u128 {
+        _mm_extract_epi64(v, I) as u128
+    }
+
+    /// clmul with 32-bit inputs (and a 64-bit answer).
+    pub unsafe fn clmul32(a: u64, b: u64) -> u128 {
+        extract::<0>(clmul(a, b))
+    }
+
+    /// clmul with 64-bit inputs (and a 128-bit answer).
+    pub unsafe fn clmul64(a: u64, b: u64) -> u128 {
+        let product = clmul(a, b);
+        extract::<1>(product) << 64 | extract::<0>(product)
+    }
+}
+
+#[allow(unreachable_code)]
+fn clmul<const BITS: u32>(a: u64, b: u64) -> u128 {
+    #[cfg(all(target_arch = "x86_64", target_feature = "pclmulqdq"))]
+    return if BITS <= 32 {
+        unsafe { clmul_x86_64::clmul32(a, b) }
+    } else if BITS <= 64 {
+        unsafe { clmul_x86_64::clmul64(a, b) }
+    } else {
+        unreachable!();
+    };
+
+    debug_assert!(2 * BITS <= u128::BITS);
```

Comment:
it is a little worrying that we won't have coverage here in CI but it does not seem easy to add :(
maybe one day we would build a docker image that emulates ARM and test IPA there

Reply:
Yeah, getting access to aarch64 hardware is hard.  We could emulate, but I don't know if the emulator has these features.